src/main/java/bio/terra/service/dataset/DatasetDao.java

```diff
@@ -483,6 +483,11 @@ public boolean delete(UUID id) {
     return rowsAffected > 0;
   }
 
+  public Dataset retrieve(UUID id, boolean retrieveRelationship, boolean retrieveAsset) {
```

Comment:
I think it's a hazard to have a single "dataset" object whose contents may be initialized differently depending on how it's created, as we don't keep track of which type of initialization was used.

A stricter approach would either store the retrieval flags in the dataset and then throw if an invalid getter was called, or (even better) having different dataset types depending on which fields are present, so invalid access would be known at compile time. Both of these are out of scope for this PR though.

Reply:
I agree, I could see this biting us down the line while also seeing the current approach as good for now.  I'd like to talk with you more about alternative approaches here -- I had also thought about different dataset types but would appreciate a thought partner to solidify what that would look like.